src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs

```diff
@@ -124,27 +128,39 @@ private class Searcher
 #pragma warning disable CS0618 // MatchKind is obsolete
             private void ReportMatchResult(Project project, INavigateToSearchResult result)
             {
+                ImmutableArray<Span> arrayTextSpan = ImmutableArray<Span>.Empty;
+                foreach (var s in result.NameMatchSpans)
+                {
+                    arrayTextSpan.Add(new Span(s.Start, s.Length));
+                }
+
+                var patternMatch = new PatternMatch(GetMatchKind(result.MatchKind), 
+                    result.IsCaseSensitive, true, arrayTextSpan);
+
                 var navigateToItem = new NavigateToItem(
                     result.Name,
                     result.Kind,
                     GetNavigateToLanguage(project.Language),
                     result.SecondarySort,
                     result,
-                    GetMatchKind(result.MatchKind),
-                    result.IsCaseSensitive,
+                    patternMatch,
                     _displayFactory);
                 _callback.AddItem(navigateToItem);
             }
 
-            private MatchKind GetMatchKind(NavigateToMatchKind matchKind)
+            private PatternMatchKind GetMatchKind(NavigateToMatchKind matchKind)
             {
                 switch (matchKind)
                 {
-                    case NavigateToMatchKind.Exact: return MatchKind.Exact;
-                    case NavigateToMatchKind.Prefix: return MatchKind.Prefix;
-                    case NavigateToMatchKind.Substring: return MatchKind.Substring;
-                    case NavigateToMatchKind.Regular: return MatchKind.Regular;
-                    default: return MatchKind.None;
+                    case NavigateToMatchKind.Exact: return PatternMatchKind.Exact;
+                    case NavigateToMatchKind.Prefix: return PatternMatchKind.Prefix;
+                    case NavigateToMatchKind.Substring: return PatternMatchKind.Substring;
+                    case NavigateToMatchKind.CamelCaseExact: return PatternMatchKind.CamelCaseExact;
+                    case NavigateToMatchKind.CamelCasePrefix: return PatternMatchKind.CamelCasePrefix;
+                    case NavigateToMatchKind.CamelCaseNonContiguousPrefix: return PatternMatchKind.CamelCaseNonContiguousPrefix;
+                    case NavigateToMatchKind.CamelCaseSubstring: return PatternMatchKind.CamelCaseSubstring;
+                    case NavigateToMatchKind.Fuzzy: return PatternMatchKind.Fuzzy;
```

Comment:
It is possible to just remove the NavigateToMatchKind enum entirely, and just use PatternMatchKind throughout our own API?

Reply:
NavigateToMatchKind is slightly different because it has an entry for Regular, which PatternMatchKind does not have.   Didn't want to remove Regular because we believe that TypeScript is also using INavigateToSerachResult and we might break them.  So we tacked on the new kinds from PatternMatchKind to NavigateToMatchKind at the end.

Reply:
Gotcha

Reply:
I'm confused though.  Won't 'regular' now throw UnexpectedValue?

Reply:
The key problem is that INavigateToSearchResult is used by TypeScript thanks to InternalsVisibleTo. INavigateToSearchResult is also used to pass search results from portable to VS-dependent layers and it cannot reference editor's PatternMatchKind because editor API is not yet portable.
The solution we came up with was to "extend" NavigateToMatchKind enum so that Roslyn can preserve pattern match kind information, but still not break TypeScript.